lib/Dialect/ESI/cosim/cosim_dpi_server/DpiEntryPoints.cpp

```diff
@@ -0,0 +1,208 @@
+//===- DpiEntryPoints.cpp - ESI cosim DPI calls -----------------*- C++ -*-===//
+//
+// Cosim DPI function implementations. Mostly C-C++ gaskets to the C++
+// RpcServer.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/cosim/Server.h"
+#include "circt/Dialect/ESI/cosim/dpi.h"
+
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+
+using namespace circt::esi::cosim;
+
+static RpcServer *server = nullptr;
+
+// ---- Helper functions ----
+
+/// Get the TCP port on which to listen. Defaults to 0xECD (ESI Cosim DPI), 3789
+/// in decimal.
+int getPort() {
+  const char *portEnv = getenv("COSIM_PORT");
+  if (portEnv == nullptr) {
+    printf("[COSIM] RPC server port not found. Defaulting to 3789.\n");
+    return 0xECD;
+  }
+  printf("[COSIM] Opening RPC server on port %s\n", portEnv);
+  return std::strtoull(portEnv, nullptr, 10);
+}
+
+/// Check that an array is an array of bytes and has some size.
+// NOLINTNEXTLINE(misc-misplaced-const)
+static int validateSvOpenArray(const svOpenArrayHandle data,
+                               int expectedElemSize) {
+  int err = 0;
+  if (svDimensions(data) != 1) {
+    printf("DPI-C: ERROR passed array argument that doesn't have expected 1D "
+           "dimensions\n");
+    err++;
+  }
+  if (svGetArrayPtr(data) == NULL) {
+    printf("DPI-C: ERROR passed array argument that doesn't have C layout "
+           "(ptr==NULL)\n");
+    err++;
+  }
+  int totalBytes = svSizeOfArray(data);
+  if (totalBytes == 0) {
+    printf("DPI-C: ERROR passed array argument that doesn't have C layout "
+           "(total_bytes==0)\n");
+    err++;
+  }
+  int numElems = svSize(data, 1);
+  int elemSize = numElems == 0 ? 0 : (totalBytes / numElems);
+  if (numElems * expectedElemSize != totalBytes) {
+    printf("DPI-C: ERROR: passed array argument that doesn't have expected "
+           "element-size: expected=%d actual=%d\n",
+           expectedElemSize, elemSize);
+    err++;
+  }
+  return err;
```

Comment:
i'm not totally sure how the conditions here will play out. is it possible for there to be multiple ways to return `err == 2`, for example?

Reply:
i guess non-zero is all that matters for this function, so not really a concern.

Reply:
Yeah, this was intended to count the errors but now I don't see why that's useful. I'll just replace the err increments with returns.

Reply:
It may be related to my personal coding style: returns should only come at the end of functions.